// @number-flow/react/dist/index.js
"use strict";
"use client";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  NumberFlowElement: () => NumberFlowElement,
  NumberFlowGroup: () => NumberFlowGroup,
  default: () => src_default,
  useCanAnimate: () => useCanAnimate
});
module.exports = __toCommonJS(src_exports);
var React = __toESM(require("react"));
var import_number_flow = require("number-flow");
var import_esm_env = require("esm-env");
var REACT_MAJOR = parseInt(React.version.match(/^(\d+)\./)?.[1]);
var isReact19 = REACT_MAJOR >= 19;
var OBSERVED_ATTRIBUTES = ["data", "digits"];
var NumberFlowElement = class extends import_number_flow.NumberFlowLite {
  attributeChangedCallback(attr, _oldValue, newValue) {
    this[attr] = JSON.parse(newValue);
  }
};
NumberFlowElement.observedAttributes = isReact19 ? [] : OBSERVED_ATTRIBUTES;
(0, import_number_flow.define)("number-flow-react", NumberFlowElement);
var formatters = {};
var serialize = isReact19 ? (p) => p : JSON.stringify;
function splitProps(props) {
  const {
    transformTiming,
    spinTiming,
    opacityTiming,
    animated,
    respectMotionPreference,
    trend,
    continuous,
    ...rest
  } = props;
  return [
    {
      transformTiming,
      spinTiming,
      opacityTiming,
      animated,
      respectMotionPreference,
      trend,
      continuous
    },
    rest
  ];
}
var NumberFlowImpl = class extends React.Component {
  constructor(props) {
    super(props);
    this.handleRef = this.handleRef.bind(this);
  }
  // Update the non-`data` props to avoid JSON serialization
  // Data needs to be set in render still:
  updateProperties(prevProps) {
    if (!this.el) return;
    this.el.manual = !this.props.isolate;
    const [nonData] = splitProps(this.props);
    Object.entries(nonData).forEach(([k, v]) => {
      this.el[k] = v ?? NumberFlowElement.defaultProps[k];
    });
    if (prevProps?.onAnimationsStart)
      this.el.removeEventListener("animationsstart", prevProps.onAnimationsStart);
    if (this.props.onAnimationsStart)
      this.el.addEventListener("animationsstart", this.props.onAnimationsStart);
    if (prevProps?.onAnimationsFinish)
      this.el.removeEventListener("animationsfinish", prevProps.onAnimationsFinish);
    if (this.props.onAnimationsFinish)
      this.el.addEventListener("animationsfinish", this.props.onAnimationsFinish);
  }
  componentDidMount() {
    this.updateProperties();
    if (isReact19 && this.el) {
      this.el.digits = this.props.digits;
      this.el.data = this.props.data;
    }
  }
  getSnapshotBeforeUpdate(prevProps) {
    this.updateProperties(prevProps);
    if (prevProps.data !== this.props.data) {
      if (this.props.group) {
        this.props.group.willUpdate();
        return () => this.props.group?.didUpdate();
      }
      if (!this.props.isolate) {
        this.el?.willUpdate();
        return () => this.el?.didUpdate();
      }
    }
    return null;
  }
  componentDidUpdate(_, __, didUpdate) {
    didUpdate?.();
  }
  handleRef(el) {
    if (this.props.innerRef) this.props.innerRef.current = el;
    this.el = el;
  }
  render() {
    const [
      _,
      {
        innerRef,
        className,
        data,
        willChange,
        isolate,
        group,
        digits,
        onAnimationsStart,
        onAnimationsFinish,
        ...rest
      }
    ] = splitProps(this.props);
    return (
      // @ts-expect-error missing types
      /* @__PURE__ */ React.createElement(
        "number-flow-react",
        {
          ref: this.handleRef,
          "data-will-change": willChange ? "" : void 0,
          class: className,
          "aria-label": data.valueAsString,
          ...rest,
          role: "img",
          dangerouslySetInnerHTML: { __html: import_esm_env.BROWSER ? "" : (0, import_number_flow.renderInnerHTML)(data) },
          suppressHydrationWarning: true,
          digits: serialize(digits),
          data: serialize(data)
        }
      )
    );
  }
};
var NumberFlow = React.forwardRef(function NumberFlow2({ value, locales, format, prefix, suffix, ...props }, _ref) {
  React.useImperativeHandle(_ref, () => ref.current, []);
  const ref = React.useRef();
  const group = React.useContext(NumberFlowGroupContext);
  group?.useRegister(ref);
  const localesString = React.useMemo(() => locales ? JSON.stringify(locales) : "", [locales]);
  const formatString = React.useMemo(() => format ? JSON.stringify(format) : "", [format]);
  const data = React.useMemo(() => {
    const formatter = formatters[`${localesString}:${formatString}`] ??= new Intl.NumberFormat(
      locales,
      format
    );
    return (0, import_number_flow.formatToData)(value, formatter, prefix, suffix);
  }, [value, localesString, formatString, prefix, suffix]);
  return /* @__PURE__ */ React.createElement(NumberFlowImpl, { ...props, group, data, innerRef: ref });
});
var src_default = NumberFlow;
var NumberFlowGroupContext = React.createContext(void 0);
function NumberFlowGroup({ children }) {
  const flows = React.useRef(/* @__PURE__ */ new Set());
  const updating = React.useRef(false);
  const pending = React.useRef(/* @__PURE__ */ new WeakMap());
  const value = React.useMemo(
    () => ({
      useRegister(ref) {
        React.useEffect(() => {
          flows.current.add(ref);
          return () => {
            flows.current.delete(ref);
          };
        }, []);
      },
      willUpdate() {
        if (updating.current) return;
        updating.current = true;
        flows.current.forEach((ref) => {
          const f = ref.current;
          if (!f || !f.created) return;
          f.willUpdate();
          pending.current.set(f, true);
        });
      },
      didUpdate() {
        flows.current.forEach((ref) => {
          const f = ref.current;
          if (!f || !pending.current.get(f)) return;
          f.didUpdate();
          pending.current.delete(f);
        });
        updating.current = false;
      }
    }),
    []
  );
  return /* @__PURE__ */ React.createElement(NumberFlowGroupContext.Provider, { value }, children);
}
function useCanAnimate({ respectMotionPreference = true } = {}) {
  const [canAnimate, setCanAnimate] = React.useState(false);
  const [reducedMotion, setReducedMotion] = React.useState(false);
  React.useEffect(() => {
    setCanAnimate(import_number_flow.canAnimate);
    setReducedMotion(import_number_flow.prefersReducedMotion?.matches ?? false);
  }, []);
  React.useEffect(() => {
    if (!respectMotionPreference) return;
    const onChange = ({ matches }) => {
      setReducedMotion(matches);
    };
    import_number_flow.prefersReducedMotion?.addEventListener("change", onChange);
    return () => {
      import_number_flow.prefersReducedMotion?.removeEventListener("change", onChange);
    };
  }, [respectMotionPreference]);
  return canAnimate && (!respectMotionPreference || !reducedMotion);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NumberFlowElement,
  NumberFlowGroup,
  useCanAnimate
});
